src/lib/comms/sol-oic-server.c

```diff
@@ -613,54 +638,183 @@ sol_oic_server_unref(void)
+    int (*handle_fn)(void *data, struct sol_oic_request *request),
     struct sol_coap_server *server, struct sol_coap_packet *req,
     const struct sol_network_link_addr *cliaddr,
     struct sol_oic_server_resource *res, bool expect_payload)
 {
-    struct sol_coap_packet *response;
-    struct sol_oic_map_reader input, *input_ptr = NULL;
-    struct sol_oic_map_writer output;
+    struct sol_coap_packet *response_pkt = NULL;
+    struct sol_oic_request *request = NULL;
     sol_coap_responsecode_t code = SOL_COAP_RSPCODE_INTERNAL_ERROR;
     CborParser parser;
+    int r;
 
     OIC_SERVER_CHECK(-ENOTCONN);
 
-    response = sol_coap_packet_new(req);
-    if (!response) {
-        SOL_WRN("Could not build response packet.");
-        return -1;
-    }
-
     if (!handle_fn) {
         code = SOL_COAP_RSPCODE_NOT_IMPLEMENTED;
-        goto done;
+        goto error;
     }
 
+    request = oic_request_new(cliaddr, server, req);
+    SOL_NULL_CHECK_GOTO(request, error);
+
     if (expect_payload) {
         if (!sol_oic_pkt_has_cbor_content(req)) {
             code = SOL_COAP_RSPCODE_BAD_REQUEST;
-            goto done;
+            goto error;
         }
-        if (sol_oic_packet_cbor_extract_repr_map(req, &parser, (CborValue *)&input) != CborNoError) {
+        if (sol_oic_packet_cbor_extract_repr_map(req, &parser,
+            (CborValue *)&request->reader) != CborNoError) {
             code = SOL_COAP_RSPCODE_BAD_REQUEST;
-            goto done;
+            goto error;
         }
-        input_ptr = &input;
-    }
+    } else
+        ((CborValue *)&request->reader)->type = CborInvalidType;
+
+    r = handle_fn((void *)res->callback.data, request);
+    if (r < 0)
+        request_free(request);
+    return r;
+
+error:
+    request_free(request);
 
-    sol_oic_packet_cbor_create(response, &output);
-    code = handle_fn(cliaddr, res->callback.data, input_ptr, &output);
-    if (sol_oic_packet_cbor_close(response, &output) != CborNoError)
-        code = SOL_COAP_RSPCODE_INTERNAL_ERROR;
+    response_pkt = sol_coap_packet_new(req);
+    SOL_NULL_CHECK(response_pkt, -errno);
+    sol_coap_header_set_code(response_pkt, code);
+    return sol_coap_send_packet(server, response_pkt, cliaddr);
+}
+
+SOL_API struct sol_oic_map_writer *
+sol_oic_server_response_get_writer(struct sol_oic_response *response)
+{
+    SOL_NULL_CHECK(response, NULL);
 
-done:
-    sol_coap_header_set_code(response, code);
+    return &response->writer;
+}
+
+SOL_API struct sol_oic_map_reader *
+sol_oic_server_request_get_reader(struct sol_oic_request *request)
+{
+    SOL_NULL_CHECK(request, NULL);
+
+    return &request->reader;
+}
+
+SOL_API int
+sol_oic_server_send_response(struct sol_oic_request *request, struct sol_oic_response *response, sol_coap_responsecode_t code)
+{
+    struct sol_coap_packet *pkt;
+    int r = -EINVAL;
 
-    return sol_coap_send_packet(server, response, cliaddr);
+    OIC_SERVER_CHECK(-ENOTCONN);
```

Comment:
This is not right. This function is responsible to release the response memory, in this case the response memory will leak